### ISSUE TITLE
Calling function with implicit template restrictions should generate an error

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#5172]
```php
<?php

/**
 * @template T as string|null
 */
abstract class A {
    /** @var T */
    public $foo;
}

/**
 * @extends A<string>
 */
class AChild extends A {
    public function __construct() {
        $this->foo.= "hello";
    }
}

function takesA(A $a) : void {}

/**
 * @param A<string|null> $a
 */
function takesAExplicit(A $a) : void {}

$child = new AChild();
takesA($child); // should be an error
takesAExplicit($child); // is an error
```

https://psalm.dev/r/31b7cd1bd6
